validator/validator.py

```diff
@@ -0,0 +1,362 @@
+        meta_schema = schemas_bundle[meta_schema_url]
+    else:
+        meta_schema = fetch_schema(meta_schema_url)
+        schemas_bundle[meta_schema_url] = meta_schema
+
+    resolver = jsonschema.RefResolver(
+        filename,
+        schema_data,
+        handlers=get_handlers(schemas_bundle)
+    )
+
+    jsonschema.Draft4Validator.check_schema(schema_data)
+    validator = jsonschema.Draft4Validator(meta_schema, resolver=resolver)
+    validator.validate(schema_data)
+
+    try:
+        jsonschema.Draft4Validator.check_schema(schema_data)
+        validator = jsonschema.Draft4Validator(meta_schema, resolver=resolver)
+        validator.validate(schema_data)
+    except jsonschema.ValidationError as e:
+        return ValidationError(kind, filename, "VALIDATION_ERROR", e,
+                               meta_schema_url)
+    except (jsonschema.SchemaError,
+            jsonschema.exceptions.RefResolutionError) as e:
+        return ValidationError(kind, filename, "SCHEMA_ERROR", e,
+                               meta_schema_url)
+
+    return ValidationOK(kind, filename, meta_schema_url)
+
+
+def validate_file(schemas_bundle, filename, data):
+    kind = ValidatedFileKind.DATA_FILE
+
+    logging.info('validating file: {}'.format(filename))
+
+    try:
+        schema_url = data[u'$schema']
+    except KeyError as e:
+        return ValidationError(kind, filename, "MISSING_SCHEMA_URL", e)
+
+    if not schema_url.startswith('http') and not schema_url.startswith('/'):
+        schema_url = '/' + schema_url
+
+    schema = schemas_bundle[schema_url]
+
+    try:
+        resolver = jsonschema.RefResolver(
+            schema_url,
+            schema,
+            handlers=get_handlers(schemas_bundle)
+        )
+        validator = jsonschema.Draft4Validator(schema, resolver=resolver)
+        validator.validate(data)
+    except jsonschema.ValidationError as e:
+        return ValidationError(kind, filename, "VALIDATION_ERROR", e,
+                               schema_url=schema_url)
+    except jsonschema.SchemaError as e:
+        return ValidationError(kind, filename, "SCHEMA_ERROR", e,
+                               schema_url=schema_url)
+    except TypeError as e:
+        return ValidationError(kind, filename, "SCHEMA_TYPE_ERROR", e,
+                               schema_url=schema_url)
+
+    return ValidationOK(kind, filename, schema_url)
+
+
+def validate_ref(schemas_bundle, bundle, filename, data, ptr, ref):
+    kind = ValidatedFileKind.REF
+
+    try:
+        ref_data = bundle[ref["$ref"]]
+    except KeyError as e:
+        return ValidationError(
+            kind,
+            filename,
+            "FILE_NOT_FOUND",
+            e,
+            ref=ref['$ref']
+        )
+
+    schema = schemas_bundle[data['$schema']]
+    schema_info = get_schema_info_from_pointer(schema, ptr)
+    expected_schema = schema_info.get('$schemaRef')
+
+    if expected_schema and expected_schema != ref_data['$schema']:
```

Comment:
`if expected_schema is not None and ...`

Reply:
totally, thanks